### PR TITLE
Store THIR in `IndexVec`s instead of an `Arena`

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
This is a necessary step to store the THIR in a query: #85273. See [relevant discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/278509-project-thir-unsafeck/topic/THIR-dependent.20queries.20design).

r? @ghost cc @cjgillot @nikomatsakis

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/mod.rs
Comment (cjgillot) [inline compiler/rustc_mir_build/src/thir/mod.rs:383]: A `Box<[Ty<'tcx>]>` should be enough here also.

**Comment by LeSeulArtichaut** [comment]
Since #83129 landed I need to rework the THIR visitor for this to be mergeable.
In the meantime r? @nikomatsakis

File: compiler/rustc_driver/src/pretty.rs
Comment (lqd) [inline compiler/rustc_driver/src/pretty.rs:506]: seems like this file is still WIP ?

Comment (LeSeulArtichaut) [reply compiler/rustc_driver/src/pretty.rs:506]: I'd like not to block this on `-Zunpretty=thir-tree`, since landing this will unblock making THIR a query

Comment (LeSeulArtichaut) [reply compiler/rustc_driver/src/pretty.rs:506]: I should remove the code commented out though

Comment (nikomatsakis) [reply compiler/rustc_driver/src/pretty.rs:492]: can we open an issue on the project-thir-unsafeck repo and reference it here in a comment? e.g., FIXME(rustlang/project-thir-unsafeck#123)

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors delegate+

r=me with a comment =)

**Comment by Mark-Simulacrum** [comment]
The performance results here (https://perf.rust-lang.org/compare.html?start=e5f83d24aee866a14753a7cedbb4e301dfe5bef5&end=d2b73fb0baf8212d4cfb135fb5ed306e9d6e49bb) look like a regression across the board -- is there motivation for this change that explains why this is necessary? The PR description is empty and I'm not seeing too much discussion here.

Happy to kick off another perf run if we think it's been mitigated since then.

**Comment by LeSeulArtichaut** [comment]
@Mark-Simulacrum The perf impact hasn't been mitigated since the perf run you mentioned, but this allows us to create a stealable query for THIR (#85273). I've added a bit of context in the PR description.

**Comment by lqd** [comment]
Do we expect making a stealable query for THIR to alleviate the max-rss increase seen in the perf run above btw ?

**Comment by lqd** [comment]
@bors retry

looks like an intermittent failure, other PRs have had a networking issue on the same mac builder

**Pull request merged**
